pulsar-io/elastic-search/src/main/java/org/apache/pulsar/io/elasticsearch/ElasticSearchClient.java

```diff
@@ -105,114 +55,54 @@
-        this.backoffRetry = new RandomExponentialRetry(elasticSearchConfig.getMaxRetryTimeInSec());
-        if (config.isBulkEnabled() == false) {
-            bulkProcessor = null;
-        } else {
-            BulkProcessor.Builder builder = BulkProcessor.builder(
-                    (bulkRequest, bulkResponseActionListener) -> client.bulkAsync(bulkRequest, RequestOptions.DEFAULT, bulkResponseActionListener),
-                    new BulkProcessor.Listener() {
-                        @Override
-                        public void beforeBulk(long l, BulkRequest bulkRequest) {
-                        }
-
-                        @Override
-                        public void afterBulk(long l, BulkRequest bulkRequest, BulkResponse bulkResponse) {
-                            log.trace("Bulk request id={} size={}:", l, bulkRequest.requests().size());
-                            for (int i = 0; i < bulkResponse.getItems().length; i++) {
-                                DocWriteRequest<?> request = bulkRequest.requests().get(i);
-                                Record record = records.get(request);
-                                BulkItemResponse bulkItemResponse = bulkResponse.getItems()[i];
-                                if (bulkItemResponse.isFailed()) {
-                                    record.fail();
-                                    try {
-                                        hasIrrecoverableError(bulkItemResponse);
-                                    } catch(Exception e) {
-                                        log.warn("Unrecoverable error:", e);
-                                    }
-                                } else {
-                                    record.ack();
-                                }
-                                records.remove(request);
-                            }
-                        }
-
-                        @Override
-                        public void afterBulk(long l, BulkRequest bulkRequest, Throwable throwable) {
-                            log.warn("Bulk request id={} failed:", l, throwable);
-                            for (DocWriteRequest<?> request : bulkRequest.requests()) {
-                                Record record = records.remove(request);
-                                record.fail();
-                            }
-                        }
+        final BulkProcessor.Listener bulkListener = new BulkProcessor.Listener() {
+
+            private Record removeAndGetRecordForOperation(BulkProcessor.BulkOperationRequest operation) {
+                return records.remove(operation.getOperationId());
+
+            }
+            @Override
+            public void afterBulk(long executionId, List<BulkProcessor.BulkOperationRequest> bulkOperationList,
+                                  List<BulkProcessor.BulkOperationResult> results) {
+                log.trace("Bulk request id={} size={}:", executionId, bulkOperationList.size());
```

Comment:
surround with "if log.isTraceEnabled()" to avoid unnecessary boxing of the parameters